marxs/optics/multiLayerMirror.py

```diff
@@ -0,0 +1,104 @@
+import numpy as np
+from astropy.io import ascii
+
+from .base import FlatOpticalElement
+from ..math.pluecker import *
+
+
+class MultiLayerMirror(FlatOpticalElement):
+    '''Multilayer mirror with varying layer thickness along one axis
+    
+    The distance between layers changes along the local y axis.
+    All reflectivity data is assumed to be for a single, desired angle. There
+    is currently no way to enter varying reflection that depends on the angle
+    of incidence.
+    Provide reflectivity data in a file with columns:
+    	'X(mm)' - position along the "changing" axis
+    	'Peak lambda' - wavelength with maximum reflection at a given position
+    	'Peak' - maximum reflection at a given position
+    	'FWHM(nm)' - full width half max, measure of width of reflection Gaussian peaks
+    
+    Parameters
+    ----------
+    fileCode: string
+    	serial number of mirror, used to open reflection data file
+    '''
+    def __init__(self, fileCode, **kwargs):
+        self.fileName = './marxs/optics/docs/' + fileCode + '.txt'
+        if ('zoom' not in kwargs):
+        	kwargs['zoom'] = np.array([1, 24.5, 12])   # in mm
+        super(MultiLayerMirror, self).__init__(**kwargs)
+
+    def process_photons(self, photons):
+        # eliminate photons that do not reach the mirror
+        doesIntersect, intersection, local = self.intersect(photons['dir'], photons['pos'])
+        photons['probability'] *= doesIntersect
+        
+        # split polarization into parallel and other components
+        beam_dir = (photons['dir'][:,0:3]).copy()
+        for i in range(0, len(beam_dir)):
+        	beam_dir[i] /= np.linalg.norm(beam_dir[i])
+        # v_1 is parallel to plane, v_2 is third direction remaining
+        v_1 = np.cross(beam_dir, self.geometry['e_x'][0:3])
+        v_2 = np.cross(beam_dir, v_1)
```

Comment:
I don't think so. `e_x` and `beam_dir` do not have to be perpendicular, so even though `v_1` is perpendicular to both, `v_2` does not have to be parallel to `e_x`. `e_x` is just the normal to the mirror, so it is useful in calculating the parallel polarization vector (`v_1`), but it is not necessarily parallel to the second polarization vector (which is "closest" to perpendicular, but not exactly perpendicular to the mirror).